org.eclipse.scanning.malcolm.core/src/org/eclipse/scanning/malcolm/core/MalcolmDevice.java

```diff
@@ -523,32 +523,25 @@ public boolean isLocked() throws MalcolmDeviceException {
 
 	@Override
 	public DeviceState latch(long time, TimeUnit unit, final DeviceState... ignoredStates) throws MalcolmDeviceException {
-
 		try {
-
 			final CountDownLatch latch = new CountDownLatch(1);
 			final List<DeviceState>     stateContainer     = new ArrayList<>(1);
 			final List<Exception> exceptionContainer = new ArrayList<>(1);
 
 			// Make a listener to check for state and then add it and latch
-			IMalcolmListener<MalcolmMessage> stateChanger = new IMalcolmListener<MalcolmMessage>() {
-				@Override
-				public void eventPerformed(MalcolmEvent<MalcolmMessage> e) {
-					MalcolmMessage msg = e.getBean();
-					try {
-						DeviceState state = MalcolmUtil.getState(msg);
-						if (state != null) {
-							if (ignoredStates!=null && Arrays.asList(ignoredStates).contains(state)) {
-								return; // Found state that we don't want!
-							}
-						}
-						stateContainer.add(state);
-						latch.countDown();
-
-					} catch (Exception ne) {
-						exceptionContainer.add(ne);
-						latch.countDown();
+			IMalcolmListener<MalcolmMessage> stateChanger = e -> {
+				MalcolmMessage msg = e.getBean();
+				try {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Extract this nested try block into a separate method. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1141)